src/matrixmessageprocessor.ts

```diff
@@ -0,0 +1,281 @@
+            // so we wrap everything in <div> just to be sure stuff is wrapped
+            // as <div> will be un-touched anyways
+            const parsed = Parser.parse(`<div>${msg.formatted_body}</div>`, {
+                lowerCaseTagName: true,
+                pre: true,
+            // tslint:disable-next-line no-any
+            } as any);
+            reply = await this.walkNode(parsed);
+            reply = reply.replace(/\s*$/, ""); // trim off whitespace at end
+        } else {
+            reply = this.escapeDiscord(msg.body);
+        }
+
+        if (msg.msgtype === "m.emote") {
+            if (profile &&
+                profile.displayname &&
+                profile.displayname.length >= MIN_NAME_LENGTH &&
+                profile.displayname.length <= MAX_NAME_LENGTH) {
+                reply = `_${profile.displayname} ${reply}_`;
+            } else {
+                reply = `_${reply}_`;
+            }
+        }
+        return reply;
+    }
+
+    private escapeDiscord(msg: string): string {
+        if (this.opts.disableEveryone) {
+            msg = msg.replace(/@everyone/g, "@ everyone");
+        }
+        if (this.opts.disableHere) {
+            msg = msg.replace(/@here/g, "@ here");
+        }
+        msg = msg.replace(/@room/g, "@here");
+        const escapeChars = ["\\", "*", "_", "~", "`"];
+        escapeChars.forEach((char) => {
+            msg = msg.replace(new RegExp("\\" + char, "g"), "\\" + char);
+        });
+        return msg;
+    }
+
+    private parsePreContent(node: Parser.HTMLElement): string {
+        let text = node.text;
+        const match = text.match(/^<code([^>]*)>/i);
+        if (!match) {
+            if (text[0] !== "\n") {
+                text = "\n" + text;
+            }
+            return text;
+        }
+        // remove <code> opening-tag
+        text = text.substr(match[0].length);
+        // remove </code> closing tag
+        text = text.replace(/<\/code>$/i, "");
+        if (text[0] !== "\n") {
+            text = "\n" + text;
+        }
+        const language = match[1].match(/language-(\w*)/i);
+        if (language) {
+            text = language[1] + text;
+        }
+        return text;
+    }
+
+    private parseUser(id: string): string {
+        const USER_REGEX = /^@_discord_([0-9]*)/;
+        const match = id.match(USER_REGEX);
+        if (!match || !this.guild.members.get(match[1])) {
+            return "";
+        }
+        return `<@${match[1]}>`;
+    }
+
+    private parseChannel(id: string): string {
+        const CHANNEL_REGEX = /^#_discord_[0-9]*_([0-9]*)/;
+        const match = id.match(CHANNEL_REGEX);
+        if (!match || !this.guild.channels.get(match[1])) {
+            return MATRIX_TO_LINK + id;
+        }
+        return `<#${match[1]}>`;
+    }
+
+    private async parsePillContent(node: Parser.HTMLElement): Promise<string> {
+        const attrs = node.attributes;
```

Comment:
Worth mentioning that this breaks non-pilling clients and bridges.

Reply:
see comment above on mentions